File: app/src/main/java/ru/sccraft/arenacontrol/UpdateActivity.java
package ru.sccraft.arenacontrol;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.AsyncTask;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.Toast;

public class UpdateActivity extends AppCompatActivity {

    private Server сервер;
    private Fe fe;
    private Поток поток;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_update);
        setTitle(R.string.pleaseWait);
        fe = new Fe(this);
        сервер = Server.Companion.fromJSON(getIntent().getStringExtra("server"));
        if (сервер == null) {
            setResult(-1);
            finish();
        }
        поток = (Поток) getLastCustomNonConfigurationInstance();
        if (поток == null) {
            поток = new Поток(this);
        } else {
            поток.link(this);
        }
    }

    @SuppressLint("StaticFieldLeak")
    private class Поток extends AsyncTask<Server, Void, Boolean> {

        UpdateActivity активность;

        Поток (UpdateActivity activity) {
            активность = activity;
            execute(сервер);
        }

        void link(UpdateActivity activity) {
            активность = activity;
        }

        @Override
        protected Boolean doInBackground(Server... сервер1) {
            Server сервер = сервер1[0];
            Boolean b = сервер.update();
            активность.fe.saveFile(активность.сервер.получить_токен() + ".json", активность.сервер.toJSON());
            return b;
        }

        @Override
        protected void onPostExecute(Boolean aBoolean) {
            super.onPostExecute(aBoolean);
            if (!aBoolean) {
                Toast.makeText(активность.getApplicationContext(), R.string.error, Toast.LENGTH_SHORT).show();
            }
            Intent intent = new Intent();
            intent.putExtra("server", активность.сервер.toJSON());
            setResult(0, intent);
            finish();
        }
    }

    @Override
    public Object onRetainCustomNonConfigurationInstance() {
        return поток;
    }

    @Override
    public void onBackPressed() {
    }
}

File: app/src/main/java/ru/sccraft/arenacontrol/UpdateActivity.kt
package ru.sccraft.arenacontrol

import android.annotation.SuppressLint
import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.widget.Toast

import androidx.appcompat.app.AppCompatActivity

class UpdateActivity : AppCompatActivity() {

    private var сервер: Server? = null
    private var fe: Fe? = null
    private var поток: Поток? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_update)
        setTitle(R.string.pleaseWait)
        fe = Fe(this)
        сервер = Server.fromJSON(intent.getStringExtra("server")!!)
        поток = lastCustomNonConfigurationInstance as? Поток
        if (поток == null) {
            поток = Поток(this)
        } else {
            поток!!.link(this)
        }
    }

    @SuppressLint("StaticFieldLeak")
    private inner class Поток internal constructor(internal var активность: UpdateActivity) : AsyncTask<Server, Void, Boolean>() {

        init {
            execute(сервер)
        }

        internal fun link(activity: UpdateActivity) {
            активность = activity
        }

        override fun doInBackground(vararg сервер1: Server): Boolean? {
            val сервер = сервер1[0]
            val b = сервер.update()
            активность.fe!!.saveFile(активность.сервер!!.получить_токен() + ".json", активность.сервер!!.toJSON())
            return b
        }

        override fun onPostExecute(aBoolean: Boolean?) {
            super.onPostExecute(aBoolean)
            if (aBoolean != null)
                if ((!aBoolean)) {
                    Toast.makeText(активность.applicationContext, R.string.error, Toast.LENGTH_SHORT).show()
                }
            val intent = Intent()
            intent.putExtra("server", активность.сервер!!.toJSON())
            setResult(0, intent)
            finish()
        }
    }

    override fun onRetainCustomNonConfigurationInstance(): Any? {
        return поток
    }

    override fun onBackPressed() {}
}
